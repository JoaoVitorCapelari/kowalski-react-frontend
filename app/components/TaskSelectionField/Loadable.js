/**
 *
 * Asynchronously loads the component for TaskSelectionField
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
