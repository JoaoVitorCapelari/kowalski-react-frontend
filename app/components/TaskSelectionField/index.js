/**
*
* TaskSelectionField
*
*/

import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import OverlaySelect from '../../components/OverlaySelect';
import OverlaySelectGroup from '../../components/OverlaySelectGroup';
import OverlaySelectGroupHeader from '../../components/OverlaySelectGroupHeader';
import OverlaySelectOption from '../../components/OverlaySelectOption';

const SelectButton = styled.button`
  width: 100%;
  height: 50px !important;
`;

const renderOptions = (options) => (
  options.map(({label, value}) => (
    <OverlaySelectOption value={value} key={value}>{label}</OverlaySelectOption>
  ))
);

const renderGroupOptions = (optionGroups) => (
  optionGroups.map(({label, options}) => (
    <OverlaySelectGroup key={label}>
      <OverlaySelectGroupHeader>{label}</OverlaySelectGroupHeader>
      { renderOptions(options) }
    </OverlaySelectGroup>
  ))
);

function TaskSelectionField(props) {
  const {
    input: { onChange, onBlur, id, value },
    label,
    meta,
    meta: { touched, error },
    isTaskOverlaySelectOpened,
    onDismissTaskOverlaySelect,
    onSelectTaskClicked,
    optionGroups,
  } = props;

  const errorInputClasses = touched && error ? 'is-danger' : '';

  const labelHtml = label ? <label htmlFor={id} className="label">{label}</label> : '';
  let buttonLabel = "Select a Task";

  if (value) {
    const group = optionGroups.find((group) => group.options
      .find((option) => option.value === value)
    );
    const option = group.options.find((option) => option.value === value);
    buttonLabel = `${group.label} - ${option.label}`;
  }

  return (
    <div className="field">
      {labelHtml}
      <div className="control">
        <SelectButton {...meta} onClick={onSelectTaskClicked} onBlur={() => onBlur(value)} className={`button ${errorInputClasses}`}>
          {buttonLabel}
        </SelectButton>
        {touched && error && <p className="help is-danger">{error}</p>}
        <OverlaySelect
          title="Select a Task"
          opened={isTaskOverlaySelectOpened}
          onDismiss={onDismissTaskOverlaySelect}
          onOptionSelect={(value) => onBlur(value) && onChange(value)}
        >
          { renderGroupOptions(optionGroups) }
        </OverlaySelect>
      </div>
    </div>
  );
}

TaskSelectionField.propTypes = {
  input: PropTypes.any.isRequired,
  meta: PropTypes.any.isRequired,
  isTaskOverlaySelectOpened: PropTypes.bool.isRequired,
  onDismissTaskOverlaySelect: PropTypes.func,
  onSelectTaskClicked: PropTypes.func.isRequired,
  label: PropTypes.any,
  optionGroups: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.any.isRequired,
    }))
  }))
};

export default TaskSelectionField;
