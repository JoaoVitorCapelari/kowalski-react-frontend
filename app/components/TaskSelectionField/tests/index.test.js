import React from 'react';
import { mount } from 'enzyme';
import { Field, reduxForm } from 'redux-form/immutable';
import createHistory from 'history/createMemoryHistory';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';

import configureStore from '../../../configureStore';
import TaskSelectionField from '../index';
import { translationMessages } from '../../../i18n';

const initialState = {};
const history = createHistory();
const store = configureStore(initialState, history);

describe('<TaskSelectionField />', () => {
  it('renders', () => {
    const TestingForm =
      reduxForm({
        form: 'test',
      })(
        () => (
          <form>
            <Field
              name="taskId"
              id="taskId"
              label="task"
              component={TaskSelectionField}
              isTaskOverlaySelectOpened={false}
              onDismissTaskOverlaySelect={jest.fn()}
              onSelectTaskClicked={jest.fn()}
              onNewTaskSelected={jest.fn()}
              disabled={false}
              optionGroups={[
                {
                  label: 'test',
                  value: 'test',
                  options: [{
                    label: 'Test',
                    value: 'Test',
                  }],
                },
              ]}
            />
          </form>
        )
      );
    const renderedComponent = mount(
      <Provider store={store}>
        <IntlProvider locale="en" messages={translationMessages.en}>
          <TestingForm />
        </IntlProvider>
      </Provider>
    );
    expect(renderedComponent.html()).toMatchSnapshot();
  });
});
