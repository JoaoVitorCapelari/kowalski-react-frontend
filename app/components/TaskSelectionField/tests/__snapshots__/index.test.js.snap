// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskSelectionField /> renders 1`] = `"<form><div class=\\"field\\"><label class=\\"label\\">task</label><div class=\\"control\\"><button class=\\"button TaskSelectionField__SelectButton-cWKMVi gcdYcW\\" form=\\"test\\">Select a Task</button><div></div></div></div></form>"`;
